packages/taiko-client/driver/soft_blocks/api.go

```diff
@@ -0,0 +1,307 @@
+		c.Request().Context(),
+		new(big.Int).SetUint64(reqBody.TransactionBatch.BlockID),
+	)
+	if err != nil && err.Error() != ethereum.NotFound.Error() {
+		return s.returnError(c, http.StatusInternalServerError, err)
+	}
+	if l1Origin != nil {
+		if l1Origin.EndOfBlock {
+			return s.returnError(c, http.StatusBadRequest, errors.New("soft block has already been marked as ended"))
+		}
+		if l1Origin.EndOfPreconf {
+			return s.returnError(
+				c,
+				http.StatusBadRequest,
+				errors.New("preconfirmation has already been marked as ended"),
+			)
+		}
+	}
+
+	// Insert the soft block.
+	header, err := s.chainSyncer.InsertSoftBlockFromTransactionsBatch(
+		c.Request().Context(),
+		reqBody.TransactionBatch.BlockID,
+		reqBody.TransactionBatch.ID,
+		s.txListDecompressor.TryDecompress(
+			s.rpc.L2.ChainID,
+			new(big.Int).SetUint64(reqBody.TransactionBatch.BlockID),
+			reqBody.TransactionBatch.TransactionsList,
+			true,
+		),
+		reqBody.TransactionBatch.BatchMarker,
+		reqBody.TransactionBatch.BlockParams,
+	)
+	if err != nil {
+		return s.returnError(c, http.StatusInternalServerError, err)
+	}
+
+	return c.JSON(http.StatusOK, BuildSoftBlockResponseBody{BlockHeader: header})
+}
+
+// RemoveSoftBlocksRequestBody represents a request body when resetting the backend
+// L2 execution engine soft head.
+type RemoveSoftBlocksRequestBody struct {
+	// @param newLastBlockID uint64 New last block ID of the blockchain, it should
+	// @param not smaller than the canonical chain's highest block ID.
+	NewLastBlockID uint64 `json:"newLastBlockId"`
+}
+
+// RemoveSoftBlocksResponseBody represents a response body when resetting the backend
+// L2 execution engine soft head.
+type RemoveSoftBlocksResponseBody struct {
+	// @param lastBlockID uint64 Current highest block ID of the blockchain (including soft blocks)
+	LastBlockID uint64 `json:"lastBlockId"`
+	// @param lastProposedBlockID uint64 Highest block ID of the cnonical chain
+	LastProposedBlockID uint64 `json:"lastProposedBlockID"`
+	// @param headsRemoved uint64 Number of soft heads removed
+	HeadsRemoved uint64 `json:"headsRemoved"`
+}
+
+// RemoveSoftBlocks removes the backend L2 execution engine soft head.
+//
+//		@Description	 Remove all soft blocks from the blockchain beyond the specified block height,
+//	  @Description	 ensuring the latest block ID does not exceed the given height. This method will fail if
+//	  @Description	 the block with an ID one greater than the specified height is not a soft block. If the
+//	  @Description	 specified block height is greater than the latest soft block ID, the method will succeed
+//	  @Description	 without modifying the blockchain.
+//		@Param      body body RemoveSoftBlocksRequestBody true "soft blocks removing request body"
+//		@Accept			json
+//		@Produce		json
+//		@Success		200	{object} RemoveSoftBlocksResponseBody
+//		@Router			/softBlocks [delete]
+func (s *SoftBlockAPIServer) RemoveSoftBlocks(c echo.Context) error {
```

Comment:
Do we need to consider storing the transactions that belong to removed soft blocks?

Reply:
Don't think so, AVS can impl that if they need

Reply:
Do you mean the AVS node will handle the transactions?

Reply:
If they want, they can store those removed txs before sending this request